ADAL/tests/integration/legacytest/ADLegacyKeychainTokenCache.m

```diff
@@ -0,0 +1,757 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import <Security/Security.h>
+#import "ADLegacyKeychainTokenCache.h"
+#import "ADAL_Internal.h"
+#import "ADKeychainUtil.h"
+#import "ADTokenCacheItem.h"
+#import "ADTokenCacheKey.h"
+#import "ADUserInformation.h"
+#import "ADWorkplaceJoinUtil.h"
+#import "ADAuthenticationSettings.h"
+#import "ADTokenCacheItem+Internal.h"
+#import "ADHelpers.h"
+#import "MSIDTelemetryCacheEvent.h"
+#import "MSIDTelemetryEventStrings.h"
+#import "MSIDTelemetry.h"
+#import "MSIDTelemetry+Internal.h"
+
+#define KEYCHAIN_VERSION 1
+#define STRINGIFY(x) #x
+#define TOSTRING(x) STRINGIFY(x)
+#define ONE_DAY_IN_SECONDS (24*60*60)
+
+static NSString* const s_nilKey = @"CC3513A0-0E69-4B4D-97FC-DFB6C91EE132";//A special attribute to write, instead of nil/empty one.
+static NSString* const s_delimiter = @"|";
+
+static NSString* const s_libraryString = @"MSOpenTech.ADAL." TOSTRING(KEYCHAIN_VERSION);
+
+static NSString* const s_wipeLibraryString = @"Microsoft.ADAL.WipeAll." TOSTRING(KEYCHAIN_VERSION);
+
+static NSString* s_defaultKeychainGroup = @"com.microsoft.adalcache";
+static ADLegacyKeychainTokenCache* s_defaultCache = nil;
+
+@implementation ADLegacyKeychainTokenCache
+{
+    NSString* _sharedGroup;
+    NSDictionary* _default;
+}
+
++ (ADLegacyKeychainTokenCache*)defaultKeychainCache
+{
+    static dispatch_once_t s_once;
+    
+    dispatch_once(&s_once, ^{
+        s_defaultCache = [[ADLegacyKeychainTokenCache alloc] init];
+    });
+    
+    
+    return s_defaultCache;
+}
+
++ (ADLegacyKeychainTokenCache*)keychainCacheForGroup:(nullable NSString*)group
+{
+    if ([group isEqualToString:s_defaultKeychainGroup])
+    {
+        return [self defaultKeychainCache];
+    }
+    ADLegacyKeychainTokenCache* cache = [[ADLegacyKeychainTokenCache alloc] initWithGroup:group];
+    return cache;
+}
+
++ (NSString*)defaultKeychainGroup
+{
+    return s_defaultKeychainGroup;
+}
+
++ (void)setDefaultKeychainGroup:(NSString *)keychainGroup
+{
+    if (s_defaultCache)
+    {
+        MSID_LOG_ERROR(nil, @"Failed to set default keychain group, default keychain cache has already been instantiated.");
+        
+        @throw @"Attempting to change the keychain group once AuthenticationContexts have been created or the default keychain cache has been retrieved is invalid. The default keychain group should only be set once for the lifetime of an application.";
+    }
+    
+    MSID_LOG_INFO(nil, @"Setting default keychain group.");
+    MSID_LOG_INFO_PII(nil, @"Setting default keychain group to %@", keychainGroup);
+    
+    if (keychainGroup == s_defaultKeychainGroup)
```

Comment:
Seems we have a bug here -- we compare pointers instead of objects. Btw, we have same issue in MSIDKeychainTokenCache :(

Reply:
Fixed in common core: https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/87